[FLINK-19741] Allow AbstractStreamOperator to skip restoring timers if subclasses are using raw keyed state

tzulitai opened the pull request:
## What is the purpose of the change

This PR provides a temporary solution for the following issue:
- On restore, the timer services **_always_** attempts to read timers from raw keyed state streams, regardless of whether or not timers were actually written to raw keyed state. This is done to allow users to swap from heap-based timers to RocksDB timers across savepoint restores. Since there is no information about the previous timer configuration in savepoints, the only way to support the timer swapping is to always see if something was written.
- The issue is that this implementation **assumes that there are no other writers to the raw keyed streams**.
- If a `AbstractStreamOperator` implementation uses the raw keyed streams, on restore the timer services would assume that they wrote the checkpointed data and tries to read them, and then the restore obviously fails (because of read errors).

## Solution

This PR solves this by adding a flag `isUsingCustomRawKeyedState` to the `AbstractStreamOperator`, which by default is `false` so that under normal circumstances things remain as before:
```
protected boolean isUsingCustomRawKeyedState() {
    return false;
}
```

If a `AbstractStreamOperator` implementation writes to raw keyed streams, it should also override this to return `true`.
On restore, the timer services would respect this flag and skip reading from the raw keyed streams.

Note that this works due to the fact that there could only ever be one writer to raw keyed streams.
If there are multiple writers attempting to write to the raw keyed stream (i.e. legacy heap-based timers + some custom writing in user code), checkpointing would have already failed consistently, since the raw keyed stream API (`KeyedStateCheckpointOutputStream`) only allows calling `startNewKeyGroup` once for the same stream.

## Backport

This change should be backported to `release-1.11`.
Separate PR for the backport: #13762 

## Brief change log

- 10bfda6 introduces the `isUsingCustomRawKeyedState` flag to `AbstractStreamOperator`
- 2eab38b allows passing in the flag when creating the `StreamOperatorStateContext`
- c8b5f04 wires in the flag to be respected by the `StreamOperatorStateContext` instantiation, so that timer services do no read from raw keyed state if they weren't the ones who wrote to it.
- ece424e Adds a test that uses an `AbstractStreamOperator` implementation that writes to raw keyed state and verifies that snapshotting and restoring works, and the new flag is being respected. If you alter the flag to be `false` in the test, the test would fail.
- ebaacde Add an extra notice in docs mentioning that checkpoint would fail if you're using RocksDB + heap timers + writing to custom raw keyed state in some operators.

## Verifying this change

The new test `AbstractStreamOperatorTest#testCustomRawKeyedStateSnapshotAndRestore` should cover this fix.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **NO**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**YES** / no)
  - The serializers: (yes / **NO** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **NO** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**YES** / no / don't know)
  - The S3 file system connector: (yes / **NO** / don't know)

## Documentation

While technically this PR introduces a new feature, we deliberately do not mention it in the docs as raw keyed state is intended as an internal feature that users should not be using (unless in some very edge cases).

carp84 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamTaskStateInitializer.java:
```diff
@@ -57,5 +59,42 @@ StreamOperatorStateContext streamOperatorStateContext(
+	 * @return a context from which the given operator can initialize everything related to state.
+	 * @throws Exception when something went wrong while creating the context.
+	 */
+	default StreamOperatorStateContext streamOperatorStateContext(
```
Correct me if I'm wrong, but `StreamTaskStateInitializer` seems to be an internal interface rather than user-facing, so I suggest to remove this old interface and use the new one only.

There're two callers of this old interface:
* `KeyedStateInputFormat` clearly doesn't use raw keyed state
* `AbstractStreamOperatorV2` seems to be special, the javadoc marks it as some experimental new base for all stream operators, but there's no production code using it yet. I would suggest to also apply the changes to this "new base", and suggest we double check with @pnowojski about the plan.

carp84 reviewed (commented):
Thanks for the efforts @tzulitai .

The changes overall LGTM. My only concerns are about keeping the legacy method in `StreamTaskStateInitializer` and whether we should apply the same changes to `AbstractStreamOperatorV2`. Please check more details inline.

pnowojski reviewed (changes requested):
It looks like this code is missing changes in the `AbstractStreamOperatorV2`.

Additionally it looks to me like the commits should be squashed together:

- code from the first 3 commits is not tested in those commits
- code in the first 2 commits is not used/dead code upon introduction, when looking at those commits alone reader has no idea what's the context and reason for the change


So in all scenarios that I could think of in the future (someone reverting/cherry picking the change, or just looking at the code via annotations to find who, why and when introduced some change) we would be better with squashing those changes together.

tzulitai commented:
Thanks for the review @carp84 @pnowojski!
I've addressed your comments in the follow-up commits.

> Additionally it looks to me like the commits should be squashed together:

@pnowojski makes sense, I'll squash the commits when merging.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:
```diff
@@ -206,13 +207,36 @@ public final void initializeState(StreamTaskStateInitializer streamTaskStateMana
+	 * override this method to return {@code true}.
+	 *
+	 * <p>Subclasses need to explicitly indicate the use of raw keyed state because, internally,
+	 * the {@link AbstractStreamOperator} may attempt to read from it as well to restore heap-based timers and
```
`AbstractStreamOperatorV2`

and ditto two lines below.

pnowojski reviewed (commented):
I can not speak about the general idea behind this change/fix, but this solution seems a bit strange to me. If an operator is using raw keyed state, we are expecting it to override this method `isUsingCustomRawKeyedState()`? Hmmmm 🤷‍♂️ 

But I'm not blocking this (+0 for the change), as I don't have the full context.

tzulitai commented:
@pnowojski since the root cause of this issue is that the timer services are incorrectly assuming that whatever written in raw keyed state is written by them (please see details in the description of this PR), the ideal solution is to include as metadata in checkpoints / savepoints a header indicating what was used to write to raw keyed state. This way, the timer service can know to safely skip restoring from raw keyed state if it wasn't written by them (there is only ever one writer to raw keyed state streams).

However, we decided not to go with that approach because:
- Adding such a header would require some backwards compatibility path for savepoint formats
- Raw keyed state is not intended or advertised to be used by users at the moment. Moreover, if some user is really using raw keyed state right now, restoring from checkpoints would have always failed due to this issue.
- In the long term, the heap-based timers should eventually by moved to the state backends as well and no longer used raw keyed state anyways.

That's why we came up with this temporary workaround, with a flag that we expect power-users to set if they are using raw keyed state. Since the Stateful Functions project bumped into this, and this is the first time ever the issue was reported, we're expecting that StateFun is currently the only Flink user with raw keyed state and needs to set this flag.

As an alternative to the `isUsingCustomRawKeyedState()` method in this PR, I also considered a configuration flag, say `state.backend.rocksdb.migrate-timers` to provide the exact same functionality across all operators in a job.
I chose to go with `isUsingCustomRawKeyedState()` so that:
- the flag is set closer to where it is needed
- only operators that are using custom raw keyed state should skip timer restores
- otherwise, using the global config flag, _all_ operators will either try to skip or read from raw keyed state.
Either way, this is meant as an undocumented internal flag that is supposedly only used by StateFun.

banmoy commented:
I think it's better to skip snapshot timer in `StreamOperatorStateHandler` for safety when `isUsingCustomRawKeyedState()` is `true`.

pnowojski commented:
Thanks for the explanation @tzulitai. Sounds good to me.

I think @banmoy made a good suggestion. Maybe we could go even one step further? `throw new UnsupportedOperationException()` when operator with `isUsingCustomRawKeyedState()` is trying to register a timer?

tzulitai commented:
@pnowojski @banmoy 
Yes, I think it won't hurt to add checking the `isUsingCustomRawKeyedState` flag to the snapshot path of legacy raw keyed state timers.
As of now, if users are already using raw keyed state, snapshotting the timers (in the case of RocksDB + heap-based timers) would have already failed with `IOException("Key group X already registered!")`.
Additionally checking `isUsingCustomRawKeyedState` on the snapshot path gives us the opportunity to throw with a more meaningful error message.

> Maybe we could go even one step further? throw new UnsupportedOperationException() when operator with isUsingCustomRawKeyedState() is trying to register a timer?

Not quite sure about this. Registering timers should work regardless of the `isUsingCustomRawKeyedState` IF the user is not using the RocksDB + heap-based timers configuration. Technically, it is possible to do this check in the `InternalTimerServiceImpl`, I don't have a strong opinion on this. @banmoy what do you think?

pnowojski commented:
Note: I don't have a strong opinion about adding some `checkState` somewhere. It would be nice to have it, but if it's not that easy to provide it, I'm fine with the PR as it is.

banmoy commented:
It seems more reasonable to throw an exception rather than skip snapshotting timers to ensure no data loss. I also don't have a strong opinion about checking state. Now I'm fine with the PR.

tzulitai commented:
@flinkbot run azure

carp84 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java:
```diff
@@ -178,11 +180,19 @@ void snapshotState(
-				timeServiceManager.get().snapshotState(snapshotContext, operatorName);
+				final InternalTimeServiceManager<?> manager = timeServiceManager.get();
+
+				if (manager.isUsingLegacyRawKeyedStateSnapshots()) {
+					checkState(
+						!isUsingCustomRawKeyedState,
+						"Attempting to snapshot timers to raw keyed state, but this operator has custom raw keyed state to write.");
+					manager.snapshotToRawKeyedState(snapshotContext.getRawKeyedOperatorStateOutput(), operatorName);
+				}
```
It would be nice if we could have a test case to cover this.

carp84 reviewed (approved):
+1, left 2 minor comments around javadoc and test case which should not be blockers.

Thanks @tzulitai for the efforts!

banmoy commented:
LGTM

tzulitai commented:
@flinkbot run azure

tzulitai commented:
@flinkbot run azure
